### ISSUE TITLE
Forbid discussion author from deleting the first post of a discussion

### ISSUE THREAD

**Issue by matteocontrini** [flarum/core#2220]
## Feature Request

**Is your feature request related to a problem? Please describe.**

On my forum quite often users delete the first post of a discussion when they instead want to delete the whole discussion. I can tell that this is the case because they often write "I'm going to delete the discussion", but then they don't, probably because they don't find the right button or settle with the first "delete" they see.

**Describe the solution you'd like**

**Deleting the first post of a discussion shouldn't be allowed**. I don't think it makes sense, I can't think of a legitimate case in which it's a good idea to do that. And it would solve the above mentioned problem.

For context, this was proposed also in #1283 by Franz. It might also be related to the implementation of #90, which is however quite old so I don't know if that code is actually still around.

**Comment by timas130** [comment]
Agree, deleting the first post in a discussion should delete the whole thread.

**Comment by clarkwinkelmann** [comment]
I think the easiest solution is to go with the disallow deleting first post, not syncing the delete between one and the other.

I realize this used to be implemented in 64e5d505332c6ba1569bbb9140107b1afd19e334 though it wasn't a proper solution as this was client-side.

We could add a special check for post number 1 in `PostPolicy::hide`:

https://github.com/flarum/core/blob/24b16f9d7cc01a3d61708358ab597adcd864fa8f/src/Post/PostPolicy.php#L139-L142

Though there's next the question of whether we want admins to still be able to delete the post number 1.

Also in some cases it might be possible for the first post of a discussion to no longer be number 1 if posts were split or merged through an extension. But I don't think it's something we need to handle in core. Preventing hiding post number 1 would solve most of the issues already.

**Comment by matteocontrini** [comment]
> I realize this used to be implemented in [64e5d50](https://github.com/flarum/core/commit/64e5d505332c6ba1569bbb9140107b1afd19e334) though it wasn't a proper solution as this was client-side.

Was it? I see server-side code doing the check in the commit you linked (in `Post.php` and `CommentPost.php`).

**Comment by clarkwinkelmann** [comment]
Indeed, I missed that. I don't have the time to dig deep enough to find what happened to that code though :sweat_smile: probably ended up removed in one of the massive refactoring like so many thing.

**Comment by matteocontrini** [comment]
Still an issue

**Comment by matteocontrini** [comment]
Still an issue

**Comment by matteocontrini** [comment]
Still an issue

**Comment by luceos** [comment]
A better solution would be to do something similar as the previous implementation. But instead of only checking for post.number = 1 also check whether the post.discussion.hidden_at is set. So that if you want to hard delete an entire discussion no unnecessary exception is thrown for the OP.

So something like:

```php
if ($post->number === 1 && ! $post->discussion->hidden_at) throw;
```
